source/retryable-writes/tests/README.rst

```diff
@@ -116,3 +120,46 @@ Use as Integration Tests
 Running these as integration tests will require a running mongod server. Each of
 these tests is valid against a standalone mongod, a replica set, and a sharded
 system for server version 3.6.0 or later.
+
+Replica Set Failover Test
+=========================
+
+In addition to network errors, writes should also be retried in the event of a
+primary failover, which results in a "not master" command error (or similar).
+The ``stepdownHangBeforePerformingPostMemberStateUpdateActions`` fail point
+implemented in `d4eb562`_ for `SERVER-31355`_ may be used for this test, as it
+allows a primary to keep its client connections open after a step down.
+
+.. _d4eb562: https://github.com/mongodb/mongo/commit/d4eb562ac63717904f24de4a22e395070687bc62
+.. _SERVER-31355: https://jira.mongodb.org/browse/SERVER-31355
+
+The following test requires two MongoClient instances. An administrative client
+will connect directly to the initial primary and be used to configure the fail
+point and step down the primary. The client under test will connect to the
+replica set and be used to execute write operations.
+
+In order to guarantee that the client under test does not detect the stepped
+down primary's state change via SDAM, it must be configured with a large
+`heartbeatFrequencyMS`_ value (e.g. 60 seconds). Single-threaded drivers may
+also need to set `serverSelectionTryOnce`_ to ``false`` to ensure that server
+selection for the retry attempt waits until a new primary is elected.
+
+.. _heartbeatFrequencyMS: https://github.com/mongodb/specifications/blob/master/source/server-discovery-and-monitoring/server-discovery-and-monitoring.rst#heartbeatfrequencyms
+.. _serverSelectionTryOnce: https://github.com/mongodb/specifications/blob/master/source/server-selection/server-selection.rst#serverselectiontryonce
+
+The test proceeds as follows:
+
+- Using the client under test, insert a document and observe a successful write
+  result. This will ensure that initial discovery takes place.
+- Using the administrative client, activate the fail point by setting ``mode``
+  to ``"alwaysOn"``.
+- Using the administrative client, execute step down the primary by executing
+  the command ``{ replSetStepDown: 60, force: true}`` on the primary. Note that
+  ``replSetStepDown`` drops the connection instead of returning a command result
+  if successful (that outcome may be asserted).
```

Comment:
It drops the connection even with the failpoint set? I'd say, in either case, just wrap the whole stepdown operation in a try / catch and ignore the error, if there is one. No need to assert that the connection is dropped, we're not testing that behavior here.

Reply:
> It drops the connection even with the failpoint set?

That's my understanding based on what I saw in https://github.com/mongodb/mongo/commit/d4eb562ac63717904f24de4a22e395070687bc62

I reworded thisto  say that "any ensuing network error should be ignored".

Reply:
@ajdavis, in the `jstests/noPassthrough/replica_set_connection_stepdown.js` test that @jmikola linked to, the failpoint is eventually unset which causes the "replSetStepDown" command to complete and close all client connections.

> any ensuing network error should be ignored

You could consider saying "any ensuing network error after deactivating the fail point should be ignored" to clarify this point.

Reply:
@visemet: It sounds like I wrongfully assumed that the network error from `replSetStepDown` was the result of the command itself and not simply a byproduct of a primary dropping its connections.

Does the fail point actually leave the `replSetStepDown` command running and the client still waiting for its response without encountering a network error? If so, it sounds like we won't be able to run this test from single thread (assuming a non-async driver)? I'd want to add some more language to advise drivers to create a separate execution context if needed.

Reply:
> Does the fail point actually leave the `replSetStepDown` command running and the client still waiting for its response without encountering a network error? If so, it sounds like we won't be able to run this test from single thread (assuming a non-async driver)?

Correct, you'd need two threads of execution - one to cause the node to step down using the "administrative client" and another to perform the write operations using the client under test.

Reply:
@visemet: Thanks. Based on that, it seems like drivers will also need three clients. The third client will need to be a direct connection to the initial primary; however, that client may be used from the main execution context (where we do inserts) since `configureFailPoint` won't block.